src/main/java/com/conveyal/datatools/manager/controllers/api/LabelController.java

```diff
@@ -0,0 +1,219 @@
+        } catch (Exception e) {
+            logMessageAndHalt(req, 500, "Unknown error encountered creating label", e);
+            return null;
+        }
+    }
+
+    /**
+     * Check that updated or new label object is valid. This method should be called before a label is
+     * persisted to the database.
+     */
+    private static void validate(Request req, Label label) {
+        List<String> validationIssues = new ArrayList<>();
+        // Label is quite forgiving (sets defaults if null) and the boolean value is type checked,
+        // so there is little to validate.
+        if (StringUtils.isEmpty(label.name)) {
+            validationIssues.add("Name field must not be empty.");
+        }
+        if (!validationIssues.isEmpty()) {
+            logMessageAndHalt(
+                    req,
+                    HttpStatus.BAD_REQUEST_400,
+                    "Request was invalid for the following reasons: " + String.join(", ", validationIssues)
+            );
+        }
+    }
+
+    /**
+     * Spark HTTP endpoint to update a label. Requires that the JSON body represent all
+     * fields the updated label should contain. See a similar note on FeedSourceController for more info
+     */
+    private static Label updateLabel(Request req, Response res) throws IOException {
+        String labelId = req.params("id");
+        Label formerLabel = requestLabelById(req, Actions.MANAGE);
+        Label updatedLabel = getPOJOFromRequestBody(req, Label.class);
+
+        // Some things shouldn't be updated
+        updatedLabel.projectId = formerLabel.projectId;
+        updatedLabel.id = formerLabel.id;
+
+        validate(req, updatedLabel);
+
+        Persistence.labels.replace(labelId, updatedLabel);
+
+        return updatedLabel;
+    }
+    /**
+     * Spark HTTP endpoint to delete a label
+     *
+     * FIXME: Should this just set a "deleted" flag instead of removing from the database entirely?
+     */
+    private static Label deleteLabel(Request req, Response res) {
+        Label label = requestLabelById(req, Actions.MANAGE);
+
+        // More specific error message if label doesn't exist
+        if (label == null) {
+            logMessageAndHalt(req, 403, "Label does not exist.");
+            return null;
+        }
+
+        try {
+            // Iterate over feed sources and remove any references to the label
+            Persistence.feedSources.getAll().forEach(feedSource -> {
+                List<String> newLabels = feedSource.labels.stream()
+                        .filter(l -> !l.equals(label.id))
+                        .collect(Collectors.toList());
+                // If there are changes, update the feed source
+                if (!newLabels.equals(feedSource.labels)) {
+                    FeedSource fs = Persistence.feedSources.getById(feedSource.id);
+                    fs.labels = newLabels;
+                    Persistence.feedSources.replace(fs.id, fs);
+                }
+            });
```

Comment:
This is one of the big reasons I'm in the camp of `Label#feedSourceIds`. This involves potentially dozens of individual database updates to remove the label from a feed source. That is risky and inefficient to be making so many requests because if one of them fails, then there could be a situation where a bunch of the records still have the label entry in them. As an aside, the first statement that fetches all feed sources does so regardless of the project associated with the label.

Reply:
I spent some time thinking about how to refactor this but am starting to come around to your side. The only other option would be to ignore non-existent labels, but that doesn't seem very smart.

Reply:
I take back my coming around to your side -- if we delete a feed source when doing `Label#feedSourceIds` we have the same problem and the same number of db calls